README.md
# auto-nsc2e.sh
Author: Jeff Sani</br>
Contributors: Matt Drown</br>
Last Update: 7/11/2022</br>
Current Version: 1.2</br>

<img src="nsc2e.png" style="display:block; margin-left: auto; margin-right: auto;">
<strong>Description</strong></br>
This script automates data processing of all newnslog archives which is normally a tedious, manual process if you have many newnslog archives to process (normally up to 100 which is the default in nsagg.conf).  This script will automate the use of the nsc2e utility against a list of Citrix ADC devices and iterate through the current and archived newnslog files, process these in accord with the counters specified, and combine them to a single tab separated values file which is then downloaded to your script host for further processing.  nsc2e is an application which converts newnslog counter databases to a spreadsheet compatible tab delimited format so that data analysis can be performed more in-depthly.  It was developed back in the day by one of the orginal NetScaler devs - Jeff Xu.  Jeff if you are still out there, I hope this breathes new life into your much appreciated efforts.  nsc2e requires a conf file to indicate what specific counters should be processed as well as the target newnslog file to precess.  You can also specify filters for these if you desire to exclude data such as only data on a specific interface.  Some example conf files are included in the <a href="nsc2e/config-examples">nsc2e/config-examples</a> folder within this repo.  <p> ADC counter data is useful for a wide variety of applications such as historical usage, trending, or sizing.  The other benefit of this script is that you can schedule it to run on an interval so that you have historical data for each device stored in a central location.  By default, ADCs will purge newnslog archives once the file count is > 100. On a busy box, this might mean you only have a few days worth of sample data so you should adjust the script runtime  frequency accordingly if you want to get a full snapshot of data over a specific time interval.  A cron job creation option is available in the init script for the automated setup. If you were looking for information that would be useful for sizing a pooled license or SDX platform resource allocation, you might be interested in looking at the following set of counters for example:
</br></br>
<table>
  <tr><td colspan="2"><strong>Network</strong></td></tr>
  <tr><td>Newnslog Counter Name</td><td>Description</td></tr>
  <tr><td>allnic_tot_rx_mbits</td><td>Number of megabits received across all interfaces</td></tr>
  <tr><td>allnic_tot_rx_packets</td><td>Number of packets received all interfaces</td></tr>
  <tr><td>nic_err_rl_pkt_drops</td><td>Number of packets dropped due to platform license rate limit</td></tr>
  <tr><td colspan="2"><strong>CPU</strong></td></tr>
  <tr><td>Newnslog Counter Name</td><td>Description</td></tr>
  <tr><td>avg_cpu_usage_pcnt</td><td>This counter tracks the average CPU utilization percentage</td></tr>
  <tr><td>mgmt_cpu_usage_pcnt</td><td> 	This counter tracks the management CPU utilization percentage</td></tr>
  <tr><td>packet_cpu_usage_pcnt</td><td>This counter tracks the packet CPU utilization percentage</td></tr>
  <tr><td colspan="2"><strong>Memory</strong></td></tr>
  <tr><td>Newnslog Counter Name</td><td>Description</td></tr>
  <tr><td>mem_tot_MB</td><td>This counter tracks the total Main memory available for use by packet engine (PE), in megabytes</td></tr>
  <tr><td>mem_tot_use_MB</td><td>This counter tracks the total NetScaler Memory in use, in megabytes</td></tr>
  <tr><td>mem_usage_pcnt</td><td>This counter tracks the percentage of memory utilization on NetScaler</td></tr>
  <tr><td colspan="2"><strong>SSL</strong></td></tr>
  <tr><td>Newnslog Counter Name</td><td>Description</td></tr>
  <tr><td>ssl_tot_sslInfo_TotalTxCount</td><td>This counter tracks the number of SSL transactions on the NetScaler appliance</td></tr>
  <tr><td>ssl_tot_sslInfo_TotalSessionCount</td><td>This counter tracks the number of SSL sessions on the NetScaler appliance</td></tr>
  <tr><td>ssl_tot_sslInfo_SessionNew</td><td>This counter tracks the rate of new SSL sessions on the NetScaler appliance</td></tr>
  <tr><td>ssl_tot_sslInfo_KeyEx_RSA_2048</td><td>This counter tracks the number of RSA 2048-bit key exchanges on the NetScaler appliance</td></tr>
  <tr><td>ssl_tot_sslInfo_KeyEx_RSA_4096</td><td>This counter tracks the number of RSA 4096-bit key exchanges on the NetScaler appliance</td></tr>
  <tr><td>ssl_tot_sslInfo_KeyEx_ECDHE</td><td>This counter tracks the number of ECDHE key exchanges on the NetScaler appliance</td></tr>
  <tr><td>ssl_tot_sslInfo_Auth_RSA</td><td>This counter tracks the number of RSA authentications on the NetScaler appliance</td></tr>
  <tr><td>ssl_tot_sslInfo_Cipher_AES_256</td><td>This counter tracks the number of AES 256-bit cipher encryption on the NetScaler appliance</td></tr>
  <tr><td>ssl_cur_q_size</td><td>This counter tracks the current queue size</td></tr>
  <tr><td colspan="2"><strong>HTTP</strong></td></tr>
  <tr><td>Newnslog Counter Name</td><td>Description</td></tr>
  <tr><td>http_tot_Requests</td><td>This counter tracks HTTP requests received</td></tr>
  <tr><td>http_tot_Responses</td><td>This counter tracks HTTP responses sent</td></tr>
  <tr><td>http_tot_rxRequestBytes</td><td>This counter tracks the bytes of HTTP data received</td></tr>
  <tr><td>http_tot_rxResponseBytes</td><td>This counter tracks the bytes received as response data</td></tr>
</table>

You can learn about the specific counters available on ADC <a href="https://support.citrix.com/search/#/All%20Products?ct=All%20types&searchText=adc%20counters&sortBy=Relevance&pageIndex=1" target="_blank">here</a>. Note - Be aware that the more counters you specify in the configuration file, the longer it will take to execute the data extraction, processing, and overall script runtime will increase.  It is also recommended to run this utility script during a maintenance window or non-peak usage times as there will be a consistent load applied to the ADC management core which may impact mangement access and other related dataplane functions.

<strong>Automated Setup Steps (For CentOS/Fedora or Ubuntu Linux Host)</strong></br>
<ol type="1">
   <li>Login to your host as the user you want to create the script under</li>
   <li>Install the required Linux packages or follow this prompt in the init script</li>
   <li>Clone the repo into the desired directory on your linux host:</li>
      <ul><li>git clone https://github.com/jeffsani/auto-nsc2e.git</li></ul>
   <li>cd to auto-nsc2e/scripts</li>
   <li>manually populate the adc-list.txt file with the ADCs you want to extract data from (NSIP or SNIP with Management enabled)</li>   
   <li>Run the auto-nsc2e-init.sh script</li>
   <li>Run the auto-nsc2e.sh script</li>
</ol>

<strong>Script Requirements</strong></br>
To implement this script you will need the following if you plan to implement manually and not use the init script:
<ol type="1">
  <li>A Linux host to run the script on</li>
  <li>Citrix ADC Build Version 12.1, 13.0, or 13.1 (It may work on older builds as well but I did not test those and they are EOL)</li>
  <li>A Linux host to run the script on</li>
  <li>Populate adc-list.txt with the list of the ADCs that you would like to iterate through</li>
    <ol>Entries should be input per line with IP and Port in the format: X.X.X.X:NNN</ol>
    <ol>ADC IP addresses should be accessible to script host and have management enabled (specifically SSH)</ol>
  <li>A username and password that will work for each ADC device</li>
   <li>Required Linux Packages:</li>
       <ul>
          <li>Debian/Ubuntu: git sshpass moreutils</li>
          <li>CentOS/Fedora: git sshpass more-utils</li>
       </ul>
   <li>Environment variables set for the user running the script that contain the Citrix ADC user/pass</li>
   <li>Optional cron job to schedule the script run on an optimal interval</li>
</ol>

<strong>Required Environment Variables</strong></br>
The following variables and their respective values are required at script runtime:
<ul>
   <li>NSC2E_ADC_USER=XXX</li>
   <li>NSC2E_ADC_PASSWORD=XXX</li>
</ul>

You can run the init script which will set these for you or create manually

<strong>ADC Service Account and Command Policy</strong></br>
It is optional but recommended to create a service account on ADC to use for the purposes of running this script in lieu of just using nsroot:  

<code>add system cmdPolicy auto-nsc2e_cmdpol ALLOW "^(shell).\*mkdir\\s/var/tmp/nsc2e-tmp|^(scp).\*/var/tmp/nsc2e-tmp/\*|^(shell).\*chmod\\s744\\s/var/tmp/nsc2e-tmp/nsc2e.sh\\s/var/tmp/nsc2e-tmp/nsc2e|^(shell).\*bash\\s/var/tmp/nsc2e-tmp/nsc2e.sh|^(shell).\*rm\\s-rf\\s/var/tmp/nsc2e-tmp"</code></br>
<code>add system user auto-nsc2e -timeout 1800 -maxsession 2 -allowedManagementInterface CLI</code></br>
<code>bind system user auto-nsc2e auto-nsc2e_cmdpol 100</code>
</br></br>
<strong>Note:</strong> On 12.1 systems omit the "-allowedManagementInterface CLI" parameter as that is not supported

release-notes.txt
v1.0 - 3/5/2022
- initial script version
v1.1 - 5/7/2022
- added support for 12.1 builds of ADC
v1.2 - 7/11/2022
- Optimized run time disk space utilization to limit impact to ADC hosts

scripts/domain_read_test.sh
#!/bin/bash
# Purpose: Read Comma Separated CSV File
# Author: Vivek Gite under GPL v2.0+
# ------------------------------------------
INPUT=domains.csv
OLDIFS=$IFS
IFS=','
[ ! -f $INPUT ] && { echo "$INPUT file not found"; exit 99; }
while read -r domain challenge provider keytype
do
case $provider in
"GD")
	echo "GoDaddy Domains:"
	echo "Domain : $domain"
	echo "Challenge : $challenge"
	echo "Key type : $keytype"
;;
"CF")
	echo "CloudFlare Domains:"
	echo "Domain : $domain"
	echo "Challenge : $challenge"
	echo "Key type : $keytype"
;;
esac
done < $INPUT
IFS=$OLDIFS

auto-nsc2e.sh
#!/bin/bash
# auto-nsc2e.sh
# This script will automate nsc2e to extract and convert specific newnslog counter data to TSV format across a set of ADCs

set -o pipefail

# Set local variables
WORKINGDIR="/var/tmp/nsc2e-tmp"
LOGFILE="./log/$(date '+%m%d%Y')-auto-nsc2e.log"
DATADIR="./data"
NSC2E_CONF=~/.adcrc

# Cleanup function
function do_cleanup {
echo "Searching for old logs > 30 days and removing them..."
find ./log/*.log -not -name '*auto-nsc2e-init.log' -mtime +30 -delete
echo "Searching for old data files > 180 days and removing them..."
find ./data/*.tsv.gz -mtime +180 -delete
echo "Cleanup completed..."
}

(
# Start Logging
echo "User $(whoami) started the script"
echo "Starting auto-nsc2e Log..."

# Load common variables from conf and check vars to see if one of the required environment variables is not set
. $NSC2E_CONF
if [[ -z "$NSC2E_ADC_USER" || -z "$NSC2E_ADC_PASSWORD" ]]; then
    echo "One or more of the required environment variables for the script is not set properly, exiting..."
    exit 1;
else
  # Set SSHPASS var for automation
  export SSHPASS="$NSC2E_ADC_PASSWORD"
fi

# Loop through each ADC in adc-list.txt and process newnslog data with nsc2e
INPUT="adc-list.txt"
[ ! -f $INPUT ] && { echo "$INPUT file not found..."; exit 99; }
while IFS=":" read -r NSC2E_ADC_IP NSC2E_ADC_PORT
do
  echo "Now processing ADC at $NSC2E_ADC_IP on Port $NSC2E_ADC_PORT..."
  #Create working directory
  sshpass -e ssh -q $NSC2E_ADC_USER@$NSC2E_ADC_IP -p $NSC2E_ADC_PORT "shell mkdir $WORKINGDIR" < /dev/null
  #Transfer tool and configuration to ADC
  echo "Transfering files to ADC..."
  sshpass -e scp -q -P $NSC2E_ADC_PORT ./bin/nsc2e ./bin/nsc2e.conf ./scripts/nsc2e.sh $NSC2E_ADC_USER@$NSC2E_ADC_IP:$WORKINGDIR < /dev/null
  echo "Setting execute permissions on nsc2e and nsc2e.sh..."
  sshpass -e ssh -q $NSC2E_ADC_USER@$NSC2E_ADC_IP -p $NSC2E_ADC_PORT "shell chmod 744 $WORKINGDIR/nsc2e.sh $WORKINGDIR/nsc2e" < /dev/null
  echo "Executing nsc2e remotely..."
  sshpass -e ssh -q $NSC2E_ADC_USER@$NSC2E_ADC_IP -p $NSC2E_ADC_PORT "shell bash $WORKINGDIR/nsc2e.sh" < /dev/null
  echo "Transferring data back to script host..."
  sshpass -e scp -q -P $NSC2E_ADC_PORT $NSC2E_ADC_USER@$NSC2E_ADC_IP:$WORKINGDIR/nsc2e.tsv.gz "$DATADIR/$(date '+%m%d%Y')-$NSC2E_ADC_IP.tsv.gz" < /dev/null
  echo "Removing remote files and folders..."
  sshpass -e ssh -q $NSC2E_ADC_USER@$NSC2E_ADC_IP -p $NSC2E_ADC_PORT "shell rm -rf $WORKINGDIR" < /dev/null
  echo "Done processing $NSC2E_ADC_IP..."
done < $INPUT

# Run cleanup routine
do_cleanup

echo "All done..."
>> $LOGFILE) 2>&1 | ts '[%H:%M:%S]' | tee -a $LOGFILE

scripts/nsc2e.sh
#!/bin/bash
#nsc2e.sh
#This script initiates the processng of the nsc2e utility against all newnslog archives
#and concatenates the results into a single data file

set -o pipefail

#Change the working directory
cd /var/tmp/nsc2e-tmp

#Copy current newnslog file to working directory
cp -r /var/nslog/newnslog .

#Untar all newnslog archives
for file in /var/nslog/*.tar.gz; do tar -xzf "$file"
   find newnslog* -prune -type d | while IFS= read -r d; do 
      #Process the counter data for nCores
      for f in $d/*; do 
         ./nsc2e -c /netscaler/nsconmsg -K "$f" -f nsc2e.conf
      done
      #Concatenate the resultant files
      #head -n +1 nsc2e-nsc2e.conf-newnslog.ppe.0 >$d.tsv
      #tail -q -n +2 nsc2e-nsc2e.conf-newnslog.ppe.* >>$d.tsv
      awk 'FNR==1 && NR!=1 { while (/^"UTC"/) getline; } 1 {print}' nsc2e-nsc2e.conf-newnslog.ppe.* >$d.tsv
      #Remove ppe files
      rm nsc2e-nsc2e.conf-newnslog.ppe.*
   done
   #Remove processed newnslog file to optimize disk space usage
   rm "$file"
done

#Concatenate all data to single file
awk 'FNR==1 && NR!=1 { while (/^"UTC"/) getline; } 1 {print}' *.tsv >nsc2e.tsv

#Compress file for transfer to script host
tar -czf nsc2e.tsv.gz nsc2e.tsv

#Remove aggregate newnslog to optimize disk space usage
rm nsc2e.tsv 

auto-nsc2e-init.sh
#!/bin/bash
# auto-nsc2e-init.sh
# This script will setup your host for the auto-nsc2e script on a debian or fedora/centos based host

set -o pipefail

# Set local variables
LOGFILE="./log/$(date '+%m%d%Y')-auto-nsc2e-init.log"
NSC2E_CONF=~/.adcrc

(
# Create data and log directories and conf if they do not already exist
echo "checking for log and data directories and creating if they do not exist..."
[ ! -d "./log" ] && mkdir log; [ ! -d "./data" ] && mkdir data; [ ! -f "$NSC2E_CONF" ] && touch $NSC2E_CONF

# Prompt for and set variables 
echo "Setting script variables..."
echo "Enter the Citrix ADC user for the script: "; read ADC_USER
echo "Enter the Citrix ADC user password: "; read -s ADC_PASSWD
echo ""

#Load common variables from conf and check vars
. $NSC2E_CONF
if [[ ! -z "$NSC2E_ADC_USER" && ! -z "$NSC2E_ADC_PASSWORD" ]]; then #file exists so update vars
   echo "Exisitng variables detected - refreshing values..."
   sed -i -e "s/NSC2E_ADC_USER=.*/NSC2E_ADC_USER=$ADC_USER/" -e "s/NSC2E_ADC_PASSWORD=.*/NSC2E_ADC_PASSWORD=\'$ADC_PASSWD\'/" $NSC2E_CONF
else #file is empty so write vars
cat >>$NSC2E_CONF <<-EOF
#Start-auto-nsc2e-Vars
export NSC2E_ADC_USER="$ADC_USER"
export NSC2E_ADC_PASSWORD='$ADC_PASSWD'
#End-auto-nsc2e-Vars
EOF
fi
echo "Script variables set successfully..."

# Download and install pre-requisites
echo "Do you want to install required system pre-requisites (requires elevated privs or sudoer membership) [Y/n]? "; read ANSWER1
ANSWER1=${ANSWER1,,} # convert to lowercase
if [ "$ANSWER1" == "y" ]; then
   echo "Installing required system pre-requisites..."
   which sudo yum >/dev/null && { sudo yum install sshpass more-utils; }
   which sudo apt-get >/dev/null && { sudo apt install sshpass moreutils; }
else
   echo "Skipping install of required system pre-requisites..."
   echo "Please refer to README.md for script requirements..."
fi

# Check for existance of populated adc-list.txt and loop through each ADC in adc-list.txt and process newnslog data with nsc2e
INPUT="adc-list.txt"
[ ! -f $INPUT ] && { echo "$INPUT_FILE file not found..."; exit 99; }
if [ $(grep -cE "[0-9][0-9]*.[0-9][0-9]*\.[0-9][0-9]*.[0-9][0-9]*:[0-9][0-9]*" $INPUT) -gt 0 ]; then
   echo "At least one host found in adc-list.txt - Skipping add..."
   while IFS=: read -r NSC2E_ADC_IP NSC2E_ADC_PORT
   do
   # Check known_hosts file and presence of NSIP and add if not present
   echo "Checking for presence of $NSC2E_ADC_IP in ~/.ssh/known_hosts..."
   if [ ! -r ~/.ssh/known_hosts ]; then mkdir -p ~/.ssh; touch ~/.ssh/known_hosts; fi
   if [ $NSC2E_ADC_PORT -eq "22" ]; then
      ssh-keygen -F $NSC2E_ADC_IP -f ~/.ssh/known_hosts &> /dev/null
      if [ "$?" -ne "0" ]; then 
         # Add ADC to known_hosts
         echo "ADC IP $NSC2E_ADC_IP not present in known_hosts - Adding IP..."
         ssh-keyscan $NSC2E_ADC_IP >> ~/.ssh/known_hosts 2> /dev/null
      else
         echo "ADC IP $NSC2E_ADC_IP already present in known_hosts - Skipping add..."
      fi
   else 
      ssh-keygen -F '[$NSC2E_ADC_IP]:$NSC2E_ADC_PORT' -f ~/.ssh/known_hosts &> /dev/null
      if [ "$?" -ne "0" ]; then 
         # Add ADC to known_hosts
         echo "ADC IP $NSC2E_ADC_IP not present in known_hosts - Adding IP..."
         ssh-keyscan -p $NSC2E_ADC_PORT $NSC2E_ADC_IP >> ~/.ssh/known_hosts 2> /dev/null
      else
         echo "ADC IP $NSC2E_ADC_IP already present in known_hosts - Skipping add..."
      fi
   fi
   done < $INPUT
else
   # Prompt for first ADC IP and Port to write to adc-list.txt
   echo "No entries found in adc-list.txt - at least one host is required to run the script..."
   echo "Input an ADC IP with management enabled - either NSIP or SNIP: "; read ANSWER2
   echo "Input the ADC Port: "; read ANSWER3
   echo "$ANSWER2:$ANSWER3" > adc-list.txt
   echo "ADC $ANSWER2:$ANSWER3 added as first entry into adc-list.txt - add any additional ADC hosts in the format X.X.X.X:NN..."
fi

# Prompt user to create cron job for scheduling the script to be run at a desired interval
echo "Would you like to schedule this script to be run - [Y/n]? "; read ANSWER4
ANSWER4=${ANSWER4,,} # convert to lowercase
if [ "$ANSWER4" == "y" ]; then
   echo "Removing old cronjob if it exists..."
   crontab -l | grep -v "auto-nsc2e.sh" | crontab -
   echo "Backing up existing entries..."
   crontab -l > auto-nsc2e
   echo "What interval would you like to run the script - [Daily/Weekly/Monthly]? "; read ANSWER5
   ANSWER5=${ANSWER5:0:1}; ANSWER5=${ANSWER5,,} # get first letter and convert to lowercase
   case $ANSWER5 in
	d)
		# Day interval cron job
      echo "Creating daily cron job at 11:59PM..."
      echo "59 11 * * * $(pwd)/auto-nsc2e.sh" >> auto-nsc2e
   ;;
	w)
		# Week interval cron job
      echo "Creating weekly cron job at 11:59PM on Saturday..."
      echo "59 11 * * 6 $(pwd)/auto-nsc2e.sh" >> auto-nsc2e
	;;
   m)
		# Month Interval cron job
      echo "Creating Monthly cron job at 11:59PM on the 30th of each month..."
      echo "59 11 30 * * $(pwd)/auto-nsc2e.sh" >> auto-nsc2e
	;;
	*)
		# Unknown input
      echo "Unknown option input - Skipping Cron setup..."
      exit 1
	;;
   esac
   crontab auto-nsc2e
   rm auto-nsc2e
   echo "Successfully created new cron job..."
fi
echo "All done - Please update bin/nsc2e.conf with desired counters prior to running auto-nsc2e.sh script..."
>> $LOGFILE) 2>&1 | ts '[%H:%M:%S]' | tee -a $LOGFILE